misk-embedded/src/main/kotlin/misk/embedded/EmbeddedMisk.kt

```diff
@@ -0,0 +1,82 @@
+package misk.embedded
+
+import com.google.inject.Injector
+import okio.buffer
+import okio.sink
+import okio.source
+import java.io.File
+import java.net.URLClassLoader
+import java.util.jar.JarFile
+
+/**
+ * Loads a Misk Service in the scope of its own class loader such that it can be accessed by in
+ * another JVM project while avoiding classpath collisions on dependencies. The embedded misk
+ * service is accessed by the host project using an "embedding interface". Use this embedding
+ * interface to create a new embedded misk service:
+ * ```
+ * val embeddedServiceInstance = EmbeddedMisk.create<EmbeddedSampleService>()
+ * ```
+ *
+ * The returned instance is the single access point for the host project to call into the embedded
+ * misk service.
+ */
+object EmbeddedMisk {
+  /**
+   * This is the Application Loader and loads all host project classes and classes depended on by
+   * the embedding interface.
+   */
+  private var hostLoader: ClassLoader = EmbeddedMisk::class.java.classLoader
+
+  inline fun <reified T : Any> create() = create(T::class.java)
+
+  fun <T : Any> create(embeddingInterfaceType: Class<T>): T {
+    val classLoader = createIsolatedLoader(embeddingInterfaceType)
+    val injector = createInjector(classLoader, "${embeddingInterfaceType.name}InjectorFactory")
+
+    return injector.getInstance(embeddingInterfaceType)!!
+  }
+
+  private fun <T : Any> copyAndExtractJar(embeddingInterfaceType: Class<T>): File {
+    val extractedJarPath = File.createTempFile(embeddingInterfaceType.name, "jar")
+    extractedJarPath.deleteOnExit()
```

Comment:
If `deleteOnExit` only deletes when the JVM is terminated, do we want to memoize all of this work in case multiple instances of an `EmbeddedMisk` service are spun up sequentially in the lifetime of one JVM (e.g., in a series of integration tests)?

Reply:
Or is that taken care of by this resolution existing somewhere in an eagerly loaded singleton?

Reply:
We should extract to a deterministic location (sha256 of .jar contents?), and avoid the work of double-extraction. This is difficult!
 - need to extract to a different place, then move. Otherwise concurrently-running tests will fail because they’ll see a partially-extracted .jar
 - need to garbage collect the old extracted objects somehow.

Perhaps this as the target name:
```
  /tmp/embedded-misk/20200708-<sha>.jar
```

Then we delete `.jar` files whose file name is less than yesterday?

```
val yesterdayPrefix = formatDate(today - 48 hours)
for (file in embeddedMiskDir.list()) {
  if (file.name < yesterdayPrefix) file.delete()
}
```

Not worth doing until we’ve proven that this thing is useful for NewsWriter. But then it’ll definitely be worth doing!